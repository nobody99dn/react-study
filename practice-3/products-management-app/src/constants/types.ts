export enum Currencies {
  Dollar = '$',
  Euro = '€',
  VND = '₫'
}

export enum ProductTypes {
  Phone = 'Phone',
  Tablet = 'Tablet',
  Laptop = 'Laptop'
}

export const PRODUCT_TYPE_LIST: ProductTypes[] = [
  ProductTypes.Phone,
  ProductTypes.Tablet,
  ProductTypes.Laptop
];

export enum FilterOrderOptions {
  Asc = 'asc',
  Desc = 'desc'
}

export const ORDER_OPTIONS = [FilterOrderOptions.Asc, FilterOrderOptions.Desc];

export const localKey = 'localStates';
