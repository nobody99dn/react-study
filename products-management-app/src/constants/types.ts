export enum ButtonVariants {
  Default = 'default',
  Primary = 'primary',
  Secondary = 'secondary'
}

export enum FwType {
  Bold = 700,
  Normal = 400,
  Light = 300
}

export enum TextFieldVariants {
  Standard = 'standard',
  Outline = 'outline',
  Error = 'error'
}

export enum Currencies {
  Dollar = '$',
  Euro = '€',
  VND = '₫'
}

export enum FormVariants {
  Create = 'Create',
  Edit = 'Edit'
}

export const PRODUCT_TYPES: string[] = [ 'Phone', 'Tablet', 'Laptop']

