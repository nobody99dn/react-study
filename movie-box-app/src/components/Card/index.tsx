import Text from '@components/Text';
import Title from '@components/Title';
import { Movie } from '@models/Movie';
import Image from 'next/future/image';
import { FC, memo } from 'react';
import RatingBox from '@components/RatingBox';
import { TitleVariants } from '@common-types/title';
import { Genres } from '@common-types/movieGenreTypes';

interface CardProps {
  className?: string;
  movie: Movie;
  onClick: () => void;
}
const Card: FC<CardProps> = ({
  className = '',
  movie: { name, genres, image, rating, releaseYear },
  onClick
}) => {
  return (
    <div
      className={`inline-block bg-white-100 shadow-2xl cursor-pointer rounded overflow-hidden${
        className && ` ${className}`
      }`}
      onClick={onClick}
    >
      <div className="relative">
        <Text
          content={releaseYear}
          className="absolute bottom-5 left-3 text-gray-100"
        />
        <Image src={image} alt={name} width={275} height={350} className="" />
      </div>
      <div className="relative px-2 py-2.5">
        <div className="w-48">
          <Title
            content={name}
            className="text-ellipsis overflow-hidden whitespace-nowrap"
          />
        </div>
        <div>
          {genres.map((genre: Genres, index: number, genres: Genres[]) => (
            <>
              <Title
                variant={TitleVariants.subtitle}
                content={genre}
                key={genre}
                className="inline-block"
              />
              {index + 1 !== genres.length && (
                <span className="text-gray-200">, </span>
              )}
            </>
          ))}
        </div>
        <RatingBox value={rating} className="absolute right-7 top-1/3" />
      </div>
    </div>
  );
};

export default memo(Card);
