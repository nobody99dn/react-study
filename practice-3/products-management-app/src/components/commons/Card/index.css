.card {
  display: inline-block;
  background-color: var(--light-gray);
  border: 1px solid var(--white);
  cursor: pointer;
  width: 100%;
  height: 27.75rem;
  max-width: 15.75rem;
  border-radius: 5px;
  text-align: start;
  margin: 1rem;
}

.card:hover {
  border: 1px solid var(--dark);
  box-shadow: inset;
}

.card-body {
  padding: 0.5rem;
}

.button-wrapper {
  display: flex;
  justify-content: space-around;
  padding: 0.5rem;
}

.card-image {
  border-radius: 5px;
  object-fit: contain;
  overflow: hidden;
  height: 18rem;
}
