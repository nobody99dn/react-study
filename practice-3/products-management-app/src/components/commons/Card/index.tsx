// Libraries
import React, { useState } from 'react';

// Components
import Image from '@components/commons/Image';
import Title, { VariantTypes } from '@components/commons/Title';
import Button from '@components/commons/Button';

// Constants
import { ButtonVariants, Currencies, FwType } from '@constants/index';

// Helpers
import { currencyFormat } from '@helpers/string';

// Styles
import './index.css';

// Type
import { Product } from '@models/product';

interface CardProps {
  product: Product;
  currency: Currencies;
  onOpenModalForm: (product: Product) => void;
  onDeleteProduct: (id: string) => void;
}

const Card: React.FC<CardProps> = ({
  product,
  currency,
  onOpenModalForm,
  onDeleteProduct
}) => {
  const { name, imageUrl, price, id, type } = product || {};

  const [isDeleteLoading, setIsDeleteLoading] = useState<boolean>(false);

  const handleToggleModal = () => {
    onOpenModalForm(product);
  };

  const handleDeleteProduct = () => {
    setIsDeleteLoading(true);
    onDeleteProduct(id);
  };

  return (
    <div className='card'>
      <Image imageUrl={imageUrl} alt={name} className='card-image' />
      <div className='card-body'>
        <div className='title-wrapper'>
          <Title className='card-title' p='0.5rem 0.5rem 0 0.5rem'>
            {name}
          </Title>
        </div>
        <Title
          color='var(--dark)'
          variant={VariantTypes.Subtitle}
          fs='italic'
          p='0 0.5rem'
          size='16px'
        >
          {type}
        </Title>
        <Title variant={VariantTypes.Subtitle} fw={FwType.Bold} p='0.5rem'>
          {currencyFormat(price)}
          <span> {currency}</span>
        </Title>
        <div className='button-wrapper'>
          <Button title='Edit' onClick={handleToggleModal} />
          <Button
            variant={ButtonVariants.Secondary}
            title='Delete'
            isLoading={isDeleteLoading}
            onClick={handleDeleteProduct}
          />
        </div>
      </div>
    </div>
  );
};

export default Card;
