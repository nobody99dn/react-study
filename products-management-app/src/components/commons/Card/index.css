.card {
  cursor: pointer;
  width: 25%;
  max-width: 17.3rem;
  border: 3px;
}

.card-body {
  padding: 0.5rem;
}

.button-wrapper {
  display: flex;
  justify-content: space-around;
  padding: 0.5rem;
}
