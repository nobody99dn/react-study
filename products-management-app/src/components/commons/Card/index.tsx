// Libraries
import React, { memo, useCallback, useState } from 'react';

// Components
import Image from '@components/commons/Image';
import Title, { VariantTypes } from '@components/commons/Title';
import Button from '@components/commons/Button';
import Modal from '@components/Modal';
import Form from '@components/Form';

// Constants
import { ERROR_MESSAGES } from '@constants/messages';
import { ButtonVariants, Currencies, FormVariants, FwType, PRODUCT_TYPE_LIST } from '@constants/types';

// Helpers
import { currencyFormat } from '@helpers/string';

// Styles
import './index.css';

// Type
import { Product } from '@type/product';

// Service
import { removeProduct } from '@services/product.service';

// Store
import { deleteProduct, error as errorAction, useStore } from '@store/index';

interface CardProps {
  product: Product;
  currency: Currencies;
  handleOpenModalForm: (product: Product) => void;
}

const Card: React.FC<CardProps> = ({
  product,
  currency,
  handleOpenModalForm
}) => {
  const { name, imageUrl, price, id, type } = product || {};
  const { globalState, dispatch } = useStore();

  const [isModalShow, setIsModalShow] = useState<boolean>(false);
  const [isDeleteLoading, setIsDeleteLoading] = useState<boolean>(false);

  const handleToggleModal = useCallback(() => {
    handleOpenModalForm(product);
  }, []);

  const handleDeleteProduct = async () => {
    setIsDeleteLoading(true);
    try {
      const deletedProduct: Product = await removeProduct(id);

      if (deletedProduct instanceof Error) {
        throw new Error(ERROR_MESSAGES.SERVER_RESPONSE_ERROR);
      }

      dispatch(deleteProduct(deletedProduct.id));
    } catch (error) {
      if (error instanceof Error) {
        dispatch(errorAction(error.message));
        return;
      }
    }

    setIsDeleteLoading(false);
  };

  return (
    <div className='card'>
      <Image
        imageUrl={imageUrl}
        alt={name}
        className='card-image'
      />
      <div className='card-body'>
        <div className='title-wrapper'>
          <Title className='card-title' p='0.5rem 0.5rem 0 0.5rem'>{name}</Title>
        </div>
        <Title
          color='var(--dark)'
          variant={VariantTypes.Subtitle}
          fs='italic'
          p='0 0.5rem'
          size='16px'
        >
          {type}
        </Title>
        <Title
          variant={VariantTypes.Subtitle}
          fw={FwType.Bold}
          p='0.5rem'
        >
          {currencyFormat(price)}
          <span> {currency}</span>
        </Title>
        <div className='button-wrapper'>
          <Button
            title='Edit'
            onClick={handleToggleModal}
          />
          <Button
            variant={ButtonVariants.Secondary}
            title='Delete'
            isLoading={isDeleteLoading}
            onClick={handleDeleteProduct}
          />
        </div>
      </div>
    </div >
  );
};

export default memo(Card);
