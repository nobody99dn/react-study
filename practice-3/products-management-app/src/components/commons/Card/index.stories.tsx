// Libraries
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import Card from './index';

// Styles
import '@assets/styles/App.css';
import '@assets/styles/variables.css';
import '@assets/styles/reset.css';
import './index.css';

// Constants
import { Currencies, ProductTypes } from '@constants/index';

export default {
  title: 'Components/Card',
  component: Card
} as ComponentMeta<typeof Card>;

const Template: ComponentStory<typeof Card> = (args) => <Card {...args} />;

export const Default = Template.bind({});
Default.args = {
  product: {
    id: '1',
    name: 'iPhone 13 Promax',
    imageUrl:
      'https://cdn-dgcei.nitrocdn.com/QaFavQVnaqgHtiSsAelwGDKVguOuACXM/assets/static/optimized/rev-991ea96/wp-content/uploads/2022/04/iPhone-13-pro-max.png',
    price: 14000000,
    type: ProductTypes.Phone
  },
  currency: Currencies.VND,
  onOpenModalForm: () => {},
  onDeleteProduct: () => {}
};
