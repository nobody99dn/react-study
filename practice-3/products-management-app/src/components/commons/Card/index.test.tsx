// Libraries
import { render, act, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';

// Components
import Card, { CardProps } from './index';
import { Product } from '@models/product';
import { iPhoneImage } from '@assets/index';
import { Currencies, ProductTypes } from '@constants/types';

describe('Card component', () => {
  const mockProduct: Product = {
    id: '999',
    name: 'Product Name',
    price: 1000000,
    imageUrl: iPhoneImage,
    type: ProductTypes.Phone
  };

  const defaultProps: CardProps = {
    product: mockProduct,
    currency: Currencies.VND,
    onOpenProductDetail: () => {},
    onDeleteProduct: () => {}
  };

  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const cardWithVnd = create(<Card {...defaultProps} />).toJSON();
    expect(cardWithVnd).toMatchSnapshot();

    const cardWithEuro = create(
      <Card {...defaultProps} currency={Currencies.Euro} />
    ).toJSON();
    expect(cardWithEuro).toMatchSnapshot();

    const cardWithDollar = create(
      <Card {...defaultProps} currency={Currencies.Dollar} />
    ).toJSON();
    expect(cardWithDollar).toMatchSnapshot();
  });

  test('should be open product detail page', () => {
    const myMock = jest.fn();

    const { getByText } = render(
      <Card {...defaultProps} onOpenProductDetail={myMock} />,
      { container }
    );

    fireEvent.click(getByText('Edit'));

    expect(myMock.mock.calls.length).toEqual(1);
  });

  test('should be delete product', () => {
    const myMock = jest.fn();

    const { getByText } = render(
      <Card {...defaultProps} onDeleteProduct={myMock} />,
      { container }
    );

    fireEvent.click(getByText('Delete'));

    expect(myMock.mock.calls.length).toEqual(1);
  });
});
