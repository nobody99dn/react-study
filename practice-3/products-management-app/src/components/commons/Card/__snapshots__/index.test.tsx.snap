// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render correctly 1`] = `
<div
  className="card"
>
  <img
    alt="Product Name"
    className="image  card-image"
    onError={[Function]}
    onLoad={[Function]}
  />
  <div
    className="card-body"
  >
    <div
      className="title-wrapper"
    >
      <div
        className="title card-title"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": undefined,
            "padding": "0.5rem 0.5rem 0 0.5rem",
          }
        }
      >
        Product Name
      </div>
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": "var(--dark)",
          "fontSize": "16px",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "padding": "0 0.5rem",
        }
      }
    >
      Phone
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
          "fontStyle": "inherit",
          "fontWeight": 700,
          "padding": "0.5rem",
        }
      }
    >
      1 000 000
      <span>
         
        ₫
      </span>
    </div>
    <div
      className="button-wrapper"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Card component should render correctly 2`] = `
<div
  className="card"
>
  <img
    alt="Product Name"
    className="image  card-image"
    onError={[Function]}
    onLoad={[Function]}
  />
  <div
    className="card-body"
  >
    <div
      className="title-wrapper"
    >
      <div
        className="title card-title"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": undefined,
            "padding": "0.5rem 0.5rem 0 0.5rem",
          }
        }
      >
        Product Name
      </div>
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": "var(--dark)",
          "fontSize": "16px",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "padding": "0 0.5rem",
        }
      }
    >
      Phone
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
          "fontStyle": "inherit",
          "fontWeight": 700,
          "padding": "0.5rem",
        }
      }
    >
      1 000 000
      <span>
         
        €
      </span>
    </div>
    <div
      className="button-wrapper"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`Card component should render correctly 3`] = `
<div
  className="card"
>
  <img
    alt="Product Name"
    className="image  card-image"
    onError={[Function]}
    onLoad={[Function]}
  />
  <div
    className="card-body"
  >
    <div
      className="title-wrapper"
    >
      <div
        className="title card-title"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": undefined,
            "padding": "0.5rem 0.5rem 0 0.5rem",
          }
        }
      >
        Product Name
      </div>
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": "var(--dark)",
          "fontSize": "16px",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "padding": "0 0.5rem",
        }
      }
    >
      Phone
    </div>
    <div
      className="subtitle"
      style={
        Object {
          "color": undefined,
          "fontSize": undefined,
          "fontStyle": "inherit",
          "fontWeight": 700,
          "padding": "0.5rem",
        }
      }
    >
      1 000 000
      <span>
         
        $
      </span>
    </div>
    <div
      className="button-wrapper"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
